src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -136,14 +138,35 @@ public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader load
                     return _compilerLoadContext.LoadFromAssemblyName(assemblyName);
                 }
 
-                var pathToLoad = _loader.GetPathToLoad(assemblyPath);
-                return LoadFromAssemblyPath(pathToLoad);
+                Debug.Assert(paths.Any());
+                if (paths.Contains(assemblyPath))
+                {
+                    return LoadFromAssemblyPath(_loader.GetPathToLoad(assemblyPath));
+                }
+
+                // The assembly isn't expected to be found at 'assemblyPath',
+                // but some assembly with the same name is known to the loader.
+                foreach (var candidatePath in paths)
+                {
+                    // Note: we assume that the assembly really can be found at this path,
+                    // and that calling GetAssemblyName doesn't cause us to hold a lock on the file
+                    var candidateName = AssemblyName.GetAssemblyName(candidatePath);
+                    // checking FullName ensures that version and PublicKeyToken match exactly
+                    if (candidateName.FullName.Equals(assemblyName.FullName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return LoadFromAssemblyPath(_loader.GetPathToLoad(candidatePath));
+                    }
+                }
+
+                // None of the candidate paths contained a compatible version of the assembly.
+                return null;
```

Comment:
Think this should be some version of `throw` here. Essentially it's not going to load into the compiler ALC, it's not a dependency so this should affirmatively fail. By `returning null` here you open the potential for the Default ALC to satisfy the dependency hence which is unexpected.

Reply:
I think you're correct. I'll add a test.

Reply:
Fixed.